File: kq-projection-dotnet/kq-projection/Services/EventHandling/PieceRemovedEventHandler.cs
using KQ_Projection.Common.Enums;
using KQ_Projection.Common.Models.Events;
using KQ_Projection.Common.Models.Jigsaw;

namespace KQ_Projection.Services.EventHandling;

public class PieceRemovedEventHandler : EventHandler
{
    public override ProjectionEventResponse Handle(byte[] eventBytes, ProjectionManager projectionManager)
    {
        AggregateEvent deserializedData = Deserialize(eventBytes, typeof(PieceRemovedEventData));
        JigsawPuzzleProjection? projection = projectionManager.GetProjection(deserializedData.gameId);

        if (projection == null)
        {
            Console.WriteLine("Projection doesn't exist!");
            return OnError(deserializedData);
        }

        return projection.RemovePuzzlePiece(deserializedData);
    }

    private ProjectionEventResponse OnError(AggregateEvent deserializedData)
    {
        PieceRemovedEventData pieceRemovedEventData = (PieceRemovedEventData)deserializedData.eventData;
        return new ProjectionEventResponse
        {
            gameId = deserializedData.gameId,
            playerId = deserializedData.playerId,
            eventStreamId = deserializedData?.eventStreamId,
            eventType = deserializedData.eventType,
            timestamp = deserializedData.timestamp,
            responseMessage = ResponseMessage.FAILURE,
            responseData = new PieceAddedEventResponseData()
            {
                pieceId = pieceRemovedEventData.pieceId

            }
        };
    }
}

File: java-apps/kq-aggregate/src/main/java/com/kafkaquest/kq/aggregate/services/handlers/PieceRemovedEventHandler.java
package com.kafkaquest.kq.aggregate.services.handlers;

import com.kafkaquest.kq.aggregate.models.HandlerResult;
import com.kafkaquest.kq.aggregate.services.JigsawPuzzleGameManager;
import com.kafkaquest.kq.common.util.models.events.eventdata.PieceRemovedEventData;
import com.kafkaquest.kq.common.util.models.events.enums.EventType;
import com.kafkaquest.kq.common.util.models.kafkakeys.AggregateKey;

import java.util.List;

public class PieceRemovedEventHandler implements EventHandler {

    @Override
    public List<HandlerResult> handle(byte[] event, JigsawPuzzleGameManager manager) {
        final var consumedEvent = deserialize(event, PieceRemovedEventData.class);

        final var aggregate = manager.getAggregate(consumedEvent.getGameId());

        final var aggregateEvent = aggregate.removePuzzlePiece(consumedEvent);

        final HandlerResult handlerResult;

        if (aggregateEvent.getEventType() == EventType.EXCEPTIONAL_EVENT) {

            handlerResult = new HandlerResult(getExceptionalEventKey(), aggregateEvent);
        } else {
            handlerResult = new HandlerResult(getKey(), aggregateEvent);
        }

        return List.of(handlerResult);
    }

    public static AggregateKey getKey() {
        return new AggregateKey(EventType.PIECE_REMOVED_EVENT);
    }
}
